### PR TITLE
Add switch domain feature; fixes #74

### LINKED ISSUE #74: Add switch domain feature
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET-WEB
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Add feature to switch domain in the active session.

### Steps to Reproduce
NA

### Logs
NA 
  
### System Configuration
- COMET version:
  - [x] COMET-WEB: 0.5.0

### PR DESCRIPTION
**Pull request by justineatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
- Add feature to be able to switch domain in the application.
- Add correction in use of CDPMessageBus listeners => improve delays in session refresh

### CODE REVIEW

File: COMETwebapp/Components/ModelDashboard/ParameterDashboard.razor [hunk 1]
```diff
@@ -88,20 +94,29 @@ Copyright (c) 2022 RHEA System S.A.
     {
         this.InitializeData();
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.InitializeData();
-            this.StateHasChanged();
-        });
+        if (!Listeners.Keys.Contains("SessionEvent"))
```
Comment (samatrhea) [inline COMETwebapp/Components/ModelDashboard/ParameterDashboard.razor:97]: please refactor using `TryGetValue`  method

File: COMETwebapp/Components/ParameterTable.razor [hunk 2]
```diff
@@ -47,26 +48,44 @@ Copyright (c) 2022 RHEA System S.A.
...
     /// </summary>
     protected override void OnInitialized()
     {
         this.InitializeParameters();
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.Parameters.Clear();
-            this.InitializeParameters();
-        });
+        if (!Listeners.Keys.Contains("SessionEvent"))
```
Comment (samatrhea) [inline COMETwebapp/Components/ParameterTable.razor:63]: please refactor using `TryGetValue` method

File: COMETwebapp/Components/ParameterTable.razor [hunk 3]
```diff
@@ -47,26 +48,44 @@ Copyright (c) 2022 RHEA System S.A.
...
     }
 
     /// <summary>
     /// Initialize the <see cref="Parameters"> list
     /// </summary>
     private void InitializeParameters()
     {
-        var ListElements = ISessionAnchor.OpenIteration.Element;
-        ListElements.ForEach(element =>
+        var ListElements = ISessionAnchor.GetIteration()?.Element;
```
Comment (samatrhea) [inline COMETwebapp/Components/ParameterTable.razor:87]: please rename the variable to `elementDefinitions`, no need to put the word `List` in the name of the variable

File: COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor [hunk 4]
```diff
@@ -92,20 +93,47 @@ Copyright (c) 2022 RHEA System S.A.
...
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.SubscribedParameters = iterationService.GetParameterSubscriptions(ISessionAnchor.OpenIteration, ISessionAnchor.CurrentDomainOfExpertise);
-            this.ViewModelSubscribedParameters.Clear();
-            this.GetViewModelSubscribedParameters();
-            this.StateHasChanged();
-        });
+    protected override void OnAfterRender(bool firstRender)
+    {
+        if (!Listeners.Keys.Contains("SessionEvent"))
```
Comment (samatrhea) [inline COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor:113]: please refactor using the `TryGetValue` method

File: COMETwebapp/Pages/ModelDashboard/DataDetails.razor [hunk 5]
```diff
@@ -96,32 +97,50 @@ Copyright (c) 2022 RHEA System S.A.
...
     {  
         this.InitializeParameters();
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.InitializeParameters();
-            this.StateHasChanged();
-        });
+        if (!Listeners.Keys.Contains("SessionEvent"))
+        {
+            Listeners.Add("SessionEvent", CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
```
Comment (samatrhea) [inline COMETwebapp/Pages/ModelDashboard/DataDetails.razor:114]: please refactor using the `TryGetValue` method

File: COMETwebapp/Pages/ModelDashboard/DataDetails.razor [hunk 6]
```diff
@@ -96,32 +97,50 @@ Copyright (c) 2022 RHEA System S.A.
...
     }
 
     /// <summary>
     /// Initialize the <see cref="BuiltParameter"> list showed in the table
     /// </summary>
     private void InitializeParameters()
     {
-        if (Criteria is not null && Domain is not null && ISessionAnchor.OpenIteration != null)
+        var iteration = ISessionAnchor.GetIteration();
+        if (Criteria is not null && Domain is not null && iteration != null)
```
Comment (samatrhea) [inline COMETwebapp/Pages/ModelDashboard/DataDetails.razor:136]: please prefix the local properties with the `this` keyword, e.g. `this.Criteria`

File: COMETwebapp/Pages/ModelDashboard/ElementsDetails.razor [hunk 7]
```diff
@@ -71,31 +72,50 @@ Copyright (c) 2022 RHEA System S.A.
...
     {
         this.InitializeElements();
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.InitializeElements();
-            this.StateHasChanged();
-        });
+        if (!Listeners.Keys.Contains("Refreshing"))
+        {
+            Listeners.Add("SessionEvent", CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x =>
```
Comment (samatrhea) [inline COMETwebapp/Pages/ModelDashboard/ElementsDetails.razor:89]: please refactor using the `TryGetValue` method

File: COMETwebapp/Pages/ModelDashboard/ModelDashboard.razor [hunk 8]
```diff
@@ -83,16 +84,33 @@ Copyright (c) 2022 RHEA System S.A.
...
+    private Dictionary<string, IDisposable> Listeners = new Dictionary<string, IDisposable>();
+
     /// <summary>
     /// Force component to rerender when the session is updated
     /// </summary>
     /// <param name="firstRender">Boolean defining first render</param>
     protected override void OnAfterRender(bool firstRender)
     {
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x =>
+        if (!Listeners.Keys.Contains("Refreshing"))
```
Comment (samatrhea) [inline COMETwebapp/Pages/ModelDashboard/ModelDashboard.razor:98]: please refactor using the `TryGetValue` method

File: COMETwebapp/Shared/NavMenu.razor [hunk 9]
```diff
@@ -38,7 +38,7 @@
                 }
             </Nodes>
         </DxTreeView>
-    } else if (ISessionAnchor.IsSessionOpen && ISessionAnchor.OpenIteration != null)
+    } else if (ISessionAnchor.IsSessionOpen && ISessionAnchor.GetIteration() != null)
```
Comment (samatrhea) [inline COMETwebapp/Shared/NavMenu.razor:41]: please put the `else if` on a new line

File: COMETwebapp/Shared/Header.razor [hunk 10]
```diff
@@ -102,16 +123,40 @@
...
         {
-            IsRefreshing = true;
-            this.StateHasChanged();
-        });
-
-        CDPMessageBus.Current.Listen<SessionStateKind>().Where(x => x == SessionStateKind.UpToDate).Subscribe(x =>
+            Listeners.Add("Refreshing", CDPMessageBus.Current.Listen<SessionStateKind>().Where(x => x == SessionStateKind.Refreshing).Subscribe(x =>
+            {
+                IsRefreshing = true;
+                this.StateHasChanged();
```
Comment (samatrhea) [inline COMETwebapp/Shared/Header.razor:131]: please replace this with `this.IsRefreshing = true;`

**Review by samatrhea — CHANGES_REQUESTED** [review]

File: COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor [hunk 11]
```diff
@@ -92,20 +93,47 @@ Copyright (c) 2022 RHEA System S.A.
...
+    /// </summary>
+    private Dictionary<string, IDisposable> Listeners = new Dictionary<string, IDisposable>();
+
     /// <summary>
     /// Initialize component at first render and after session update
     /// </summary>
     protected override void OnInitialized()
     {
-        this.SubscribedParameters = iterationService.GetParameterSubscriptions(ISessionAnchor.OpenIteration, ISessionAnchor.CurrentDomainOfExpertise);
+        this.ViewModelSubscribedParameters.Clear();
```
Comment (samatrhea) [inline COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor:106]: i would wrap these calls in a private `loadData` method (or a similar name like that`)

File: COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor [hunk 12]
```diff
@@ -92,20 +93,47 @@ Copyright (c) 2022 RHEA System S.A.
...
 
-        CDPMessageBus.Current.Listen<SessionEvent>().Where(x => x.Status == SessionStatus.EndUpdate).Subscribe(x => {
-            this.SubscribedParameters = iterationService.GetParameterSubscriptions(ISessionAnchor.OpenIteration, ISessionAnchor.CurrentDomainOfExpertise);
-            this.ViewModelSubscribedParameters.Clear();
-            this.GetViewModelSubscribedParameters();
-            this.StateHasChanged();
-        });
+    protected override void OnAfterRender(bool firstRender)
+    {
+        if (this.Listeners.TryGetValue("UpToDate", out var listener))
```
Comment (samatrhea) [inline COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor:113]: I think you need to reverse the logic here `if (!this.Listeners.TryGetValue("UpToDate", out var listener))`

File: COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor [hunk 13]
```diff
@@ -92,20 +93,47 @@ Copyright (c) 2022 RHEA System S.A.
...
+        {
+            this.Listeners.Add("UpToDate", CDPMessageBus.Current.Listen<SessionStateKind>().Where(x => x == SessionStateKind.UpToDate).Subscribe(x =>
+            {
+                this.OnInitialized();
```
Comment (samatrhea) [inline COMETwebapp/Components/SubscriptionDashboard/SubscribedTable.razor:117]: i would not call `this.OnInitialized();` here, refactor what you are doing in OnInitialized and call a new private method called something like `LoadData()`

**Review by samatrhea — CHANGES_REQUESTED** [review]

**Review by alexatrhea — APPROVED** [review]
Approved pending Sam's comments.

**Review by samatrhea — APPROVED** [review]

**Pull request merged**